### ISSUE TITLE
[feature] Compile time stricture to stop duplicate variable declarations

### ISSUE THREAD

**Issue by epa** [Perl/perl5#18530]
This is legal code:
```
my $x = 5;
{ my $x = 6 }
```
Often, however, reusing the same variable name in a child scope indicates a programmer mistake, adding the `my` when it's not needed.  Unlike other mistakes with the wrong variable, this isn't caught by `use warnings` or `use strict`.  Some other programming languages with nested scopes take a stricter approach, not allowing the same name in both parent and child scopes.  It doesn't reduce the expressivity of those languages but it does make it harder to introduce subtle bugs.

I am aware that perlcritic will report this.  This feature request is for a compile time warning, which would flag the error much sooner in the development cycle.  There is already a warning for the following code:

```
use warnings;
my $x;
my $x;
    "my" variable $x masks earlier declaration in same scope at prog line 3.
```

So I suggest it should be generalized a little to

````
"my" variable $x masks earlier declaration in enclosing scope at prog line 2.
````

**Comment by karenetheridge** [comment]
I really like this idea, although I'm not sure where it could go:

- `warnings` has all warnings enabled by default (and we have discussed several times whether to change that practice, settling on "please let's not"), so to add this to warnings (perhaps as a `shadow` subcategory under `redefine`) would change the behaviour of previously-not-warning code, which could be seen as disruptive (although we could always do a cpan smoke to get an idea of the impact there)
- `strict` doesn't have subcategories, so it would need a new category (perhaps `shadow`?).. and I think strict has the same issue of defaulting to everything enabled?

I hope I am missing something obvious that would make this feature possible, while not enabling it by default.

**Comment by Grinnz** [comment]
Adding new warnings is allowed under policy and often done - this is why fatal warnings are "at your own risk". The question to me is more whether it will be more useful than annoying to all of the code which purposely shadows outer variables.

**Comment by Leont** [comment]
> The question to me is more whether it will be more useful than annoying to all of the code which purposely shadows outer variables.

Yeah, this is the main limiting factor.

**Comment by jkeenan** [comment]
> > The question to me is more whether it will be more useful than annoying to all of the code which purposely shadows outer variables.
> 
> Yeah, this is the main limiting factor.

I concur, and therefore oppose this new feature request.

**Comment by leonerd** [comment]
I wouldn't mind it as an additional "opt-in" thing, that doesn't turn up even if you

```perl
use warnings :all;
```

I suspect I might have to be its own new thing then... `use diagnostics` or `no lint` or ... somesuch. I don't have a good name for it.

**Comment by epa** [comment]
Yes, the feature request is for the warning to be available somehow, not necessarily for it to be on by default.  Personally, I would go further and fatalize it (I like to fatalize all compile time diagnostics, but not run time ones) but I recognize that not everyone is so B&D.

**Comment by Grinnz** [comment]
Yeah unfortunately there isn't a mechanism for specifically opt-in warnings, as "use warnings" is the same as "use warnings 'all'". A new pragma might be a low friction way to introduce such.

**Comment by epa** [comment]
Perhaps for future versions the set of warnings could be conditional on the declared perl version.  So `use 5.040` followed by `use warnings` would get you the set of warnings enabled in that version.

(Personally, I don't see the fuss; I see new warnings as an improvement and am pleased to see a new perl version giving me more guidance on improving my code.  But I know that a lot of people see it differently, and consider it a bug in perl that code that worked quietly before now produces log messages or output on stderr.)

**Comment by Grinnz** [comment]
The concern isn't that it would be a breaking issue, as mentioned there is already no guarantee that warnings won't change. We have to weigh the usefulness of warnings against the risk of people deciding it's better not to "use warnings". So something like this needs some opt-in mechanism beyond "use warnings 'all'" which has been an ongoing discussion.

**Comment by iabyn** [comment]
This request feels like it should be a new strict-ism rather than a
warning. But the same backcomp issues apply as for warnings.

-- 
If life gives you lemons, you'll probably develop a citric acid allergy.

**Comment by Grinnz** [comment]
#18543 may be relevant to this.
